packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmImpl.kt

```diff
@@ -112,6 +113,13 @@ public class RealmImpl private constructor(
         var realmFileCreated = false
         try {
             runBlocking {
+                // FIXME We cannot ensure exclusive access to the realm file, so for now multiple
+                //  processes/threads could potentially race each other here
+                //  https://github.com/realm/realm-core/issues/6492
+                if (configuration.assetFile != null && !fileExists(configuration.path)) {
```

Comment:
I think this is also racy in a single process. Would it make sense to put this inside a static mutex?

Reply:
Hmm. I guess we could try something like that and narrow the window for race conditions. But unless it is build into core or we lock the whole `init`-block then there is no way to guard against deleting it between it is copied from the asset and `configuration.openRealm`, in which case core will create a new file from scratch. Since people haven't seen this as an issue in realm-java, I suspect it is fine to just ensure that only one thread will try to copy it into place within a single process as suggested!?

Reply:
Yes, that was also what I ment. Sorry if that wasn't clear. But right now it looks like there is no locking in place here at all?

Reply:
Yes, that was also what I meant. Sorry if that wasn't clear. But right now it looks like there is no locking in place here at all?

Reply:
I added a global mutex around copying the asset files in place. We could try to reduce that to only block others from writing the same `configuration.path` but I would assume that it is ok to prevent multiple asset files to be copied at the same time for now, as this only happens on first initialization. My immediate thought is that any improvements should be done in core instead.